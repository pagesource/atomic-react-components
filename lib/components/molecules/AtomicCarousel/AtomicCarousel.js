// @flow
/**
 *
 * AtomicCarousel
 *
 */
import React, { useState, useEffect, useCallback } from 'react';
import styled from 'styled-components';

import styles from './AtomicCarousel.style';
import type { Props } from './types';
import List from '../List';
import ListItem from '../../atoms/ListItem';
import { ButtonVanilla } from '../../atoms/Button';

// Rendering all dots as button
const getSliderDots = (slidesCount, currentItem, setItem) => {
  const specificNode = useCallback(index => setItem(index));
  const countElements = [];

  for (let i = 1; i <= slidesCount; i += 1) {
    countElements.push(
      <ListItem key={i} className="atomic-item">
        <ButtonVanilla disabled={currentItem === i} onClick={() => specificNode(i)}>
          {i}
        </ButtonVanilla>
      </ListItem>
    );
  }

  return countElements;
};

// Calculating required number of dots
const getList = (items, setItem, currentItem, perSlideFrame) => {
  const slidesCount = Math.ceil(items.length / perSlideFrame);

  return (
    <div className="atomic-dots">
      <List>{getSliderDots(slidesCount, currentItem, setItem)}</List>
    </div>
  );
};

// Main Component
const AtomicCarousel = ({
  dots,
  navigation,
  children,
  className,
  perSlideFrame,
  prevLabel,
  nextLabel,
}: Props): Node => {
  const carouselClasses = [...className.split(','), 'atomic-carousel'].join(' ');
  const [currentSlide, setCurrentSlide] = useState(1);
  const [data, setData] = useState([]);

  const handlePrev = useCallback(() => {
    setCurrentSlide(currentSlide - 1);
  });

  const handleNext = useCallback(() => {
    setCurrentSlide(currentSlide + 1);
  });

  useEffect(() => {
    if (children !== undefined && !Array.isArray(children)) {
      setData([children]);
    } else {
      setData(children);
    }
  }, []);

  const getPerFrameItems = (content, perFrame, currentScreen) => {
    const itemsArray = [];
    const start = currentScreen === 1 ? 0 : perFrame * (currentScreen - 1);

    if (perSlideFrame > 1) {
      for (let i = start; i <= currentScreen * perFrame - 1; i += 1) {
        itemsArray.push(content[i]);
      }
    } else {
      itemsArray.push(content[currentScreen - 1]);
    }

    return itemsArray;
  };

  return (
    <>
      <article className={carouselClasses}>
        {navigation && (
          <ButtonVanilla className="previous" onClick={handlePrev} disabled={currentSlide === 1}>
            {prevLabel}
          </ButtonVanilla>
        )}
        {data.length !== 0 && getPerFrameItems(data, perSlideFrame, currentSlide)}
        {navigation && (
          <ButtonVanilla
            className="next"
            onClick={handleNext}
            disabled={Math.ceil(data.length / perSlideFrame) === currentSlide}
          >
            {nextLabel}
          </ButtonVanilla>
        )}
      </article>
      {dots && getList(data, setCurrentSlide, currentSlide, perSlideFrame)}
    </>
  );
};

AtomicCarousel.defaultProps = {
  dots: true,
  navigation: true,
  classNames: '',
  perSlideFrame: 1,
  nextLabel: 'next',
  prevLabel: 'prev',
};

export default styled(AtomicCarousel)`
  ${styles};
`;

export { AtomicCarousel as AtomicCarouselVanilla };
