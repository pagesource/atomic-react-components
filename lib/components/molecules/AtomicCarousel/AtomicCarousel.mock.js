const defaultConfig = {
  dots: true,
  navigation: true,
  perSlideFrame: 2,
};

export default defaultConfig;
