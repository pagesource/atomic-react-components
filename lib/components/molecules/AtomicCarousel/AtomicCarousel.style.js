import { css } from 'styled-components';

const styles = css`
  display: flex;
  justify-content: space-between;
  align-items: center;

  & > * + * {
    margin-top: auto;
  }

  & + .atomic-dots {
    & > ul {
      display: flex;
      align-items: center;
      justify-content: center;
      margin: 10px;
    }
  }
`;
export default styles;
