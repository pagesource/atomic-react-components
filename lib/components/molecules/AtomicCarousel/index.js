// @flow
export { default } from './AtomicCarousel';
export { AtomicCarouselVanilla } from './AtomicCarousel';
