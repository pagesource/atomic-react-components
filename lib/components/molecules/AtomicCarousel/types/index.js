// @flow

export type Props = {
  dots: Boolean,
  navigation: Boolean,
  classNames: String,
  perSlideFrame: Number,
  nextLabel: String,
  prevLabel: String,
};
