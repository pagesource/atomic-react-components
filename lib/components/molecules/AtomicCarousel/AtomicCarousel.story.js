import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, number, text } from '@storybook/addon-knobs';
// import { action } from '@storybook/addon-actions';

import defaultConfig from './AtomicCarousel.mock';
import Image from '../../atoms/Image';
import image1 from './images/1.jpg';
import image2 from './images/2.jpg';
import image3 from './images/3.jpg';
import image4 from './images/4.jpg';

// Import Styled Component to showcase variations
import AtomicCarousel, { AtomicCarouselVanilla } from '.';

storiesOf('Molecules/AtomicCarousel', module)
  .addParameters({ jest: ['AtomicCarousel', 'AtomicCarouselVanilla'] })
  .add('AtomicCarouselVanilla Knobs', () => (
    <AtomicCarouselVanilla
      {...defaultConfig}
      dots={boolean('Dots', true)}
      navigation={boolean('Navigation', true)}
      perSlideFrame={number('No. of slides per frame', 2)}
      prevLabel={text('Previous text', 'Prev')}
      nextLabel={text('Next text', 'Next')}
      heading={text('Heading Text', 'Spotlight')}
      displayHeader={boolean('Display Header', true)}
      slideTransitionTime={number('Transition Time in ms', 1500)}
      autoPlay={boolean('Auto Play', true)}
      playLabel={text('Play Label', 'Play')}
      stopLabel={text('Pause Label', 'Stop')}
    >
      <Image src={image1} alt="1.jpg" />
      <Image src={image2} alt="2.jpg" />
      <Image src={image3} alt="3.jpg" />
      <Image src={image4} alt="4.jpg" />
    </AtomicCarouselVanilla>
  ))
  .add('AtomicCarousel', () => (
    <AtomicCarousel {...defaultConfig}>
      <Image src={image1} alt="1.jpg" />
      <Image src={image2} alt="2.jpg" />
      <Image src={image3} alt="3.jpg" />
      <Image src={image4} alt="4.jpg" />
    </AtomicCarousel>
  ));
