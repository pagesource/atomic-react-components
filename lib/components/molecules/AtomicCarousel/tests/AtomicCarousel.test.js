import { shallow, mount } from 'enzyme';

import AtomicCarousel from '../index';

describe('<AtomicCarousel />', () => {
  const AtomicCarouselComponent = shallow(
    <AtomicCarousel>
      <div className="1">Test1</div>
      <div className="2">Test2</div>
      <div className="3">Test3</div>
      <div className="4">Test4</div>
    </AtomicCarousel>
  );

  it('Should display first item initially', () => {
    expect(AtomicCarouselComponent.contains(<div className="1">Test1</div>)).toBe(true);
  });

  it('render without dots', () => {
    const wrapper = shallow(
      <AtomicCarousel dots={false}>
        <h1>Test</h1>
      </AtomicCarousel>
    );
    expect(wrapper.props().dots).toBe(false);
    expect(wrapper.find('.atomic-dots').length).toBe(0);
    expect(wrapper).toMatchSnapshot();
  });

  it('render without navigation', () => {
    const wrapper = shallow(
      <AtomicCarousel navigation={false}>
        <h1>Test</h1>
      </AtomicCarousel>
    );
    expect(wrapper.props().navigation).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  test('should render correctly', () => {
    expect(AtomicCarouselComponent).toMatchSnapshot();
  });
});
