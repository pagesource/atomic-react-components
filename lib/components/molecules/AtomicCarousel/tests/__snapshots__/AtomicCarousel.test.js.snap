// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AtomicCarousel /> render without dots 1`] = `
<StyledComponent
  dots={false}
  forwardedComponent={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "AtomicCarousel-sc-18kyvc6-0",
        "isStatic": false,
        "rules": Array [
          "display:flex;justify-content:space-between;align-items:center;& > * + *{margin-top:auto;}& + .atomic-dots{& > ul{display:flex;align-items:center;justify-content:center;margin:10px;}}",
          ";",
        ],
      },
      "displayName": "AtomicCarousel",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "AtomicCarousel-sc-18kyvc6-0",
      "target": [Function],
      "toString": [Function],
      "usesTheme": false,
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    }
  }
  forwardedRef={null}
>
  <h1>
    Test
  </h1>
</StyledComponent>
`;

exports[`<AtomicCarousel /> render without navigation 1`] = `
<StyledComponent
  forwardedComponent={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "AtomicCarousel-sc-18kyvc6-0",
        "isStatic": false,
        "rules": Array [
          "display:flex;justify-content:space-between;align-items:center;& > * + *{margin-top:auto;}& + .atomic-dots{& > ul{display:flex;align-items:center;justify-content:center;margin:10px;}}",
          ";",
        ],
      },
      "displayName": "AtomicCarousel",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "AtomicCarousel-sc-18kyvc6-0",
      "target": [Function],
      "toString": [Function],
      "usesTheme": false,
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    }
  }
  forwardedRef={null}
  navigation={false}
>
  <h1>
    Test
  </h1>
</StyledComponent>
`;

exports[`<AtomicCarousel /> should render correctly 1`] = `
<StyledComponent
  forwardedComponent={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "AtomicCarousel-sc-18kyvc6-0",
        "isStatic": false,
        "rules": Array [
          "display:flex;justify-content:space-between;align-items:center;& > * + *{margin-top:auto;}& + .atomic-dots{& > ul{display:flex;align-items:center;justify-content:center;margin:10px;}}",
          ";",
        ],
      },
      "displayName": "AtomicCarousel",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "AtomicCarousel-sc-18kyvc6-0",
      "target": [Function],
      "toString": [Function],
      "usesTheme": false,
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    }
  }
  forwardedRef={null}
>
  <div
    className="1"
  >
    Test1
  </div>
  <div
    className="2"
  >
    Test2
  </div>
  <div
    className="3"
  >
    Test3
  </div>
  <div
    className="4"
  >
    Test4
  </div>
</StyledComponent>
`;
